File: src/main/java/com/vjiki/music/service/BandService.java
package com.vjiki.music.service;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import com.vjiki.music.dto.BandResponse;
import com.vjiki.music.dto.SongResponse;
import com.vjiki.music.entity.Artist;
import com.vjiki.music.pagination.CursorPageResponse;
import com.vjiki.music.pagination.LocalDateTimeIdCursorCodec;
import com.vjiki.music.repository.ArtistRepository;

@Service
public class BandService {

    /** Upper bound on search pages walked per band, mirroring the Kotlin implementation. */
    private static final int MAX_SEARCH_PAGES = 50;

    private final ArtistRepository artistRepository;
    private final SearchService searchService;

    public BandService(ArtistRepository artistRepository, SearchService searchService) {
        this.artistRepository = artistRepository;
        this.searchService = searchService;
    }

    public List<BandResponse> getBands(int limit) {
        int safeLimit = Math.min(Math.max(limit, 1), 200);
        return artistRepository.findBands(PageRequest.of(0, safeLimit)).stream()
                .map(band -> toBandResponse(band, List.of()))
                .toList();
    }

    public CursorPageResponse<BandResponse> getBandsPage(UUID userId, String name, int limit, String cursor) {
        String q = name == null ? "" : name.trim();
        if (q.isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "name is required");
        }

        int safeLimit = Math.min(Math.max(limit, 1), 200);
        Pageable pageable = PageRequest.of(0, safeLimit + 1);

        LocalDateTimeIdCursorCodec.Cursor decoded = LocalDateTimeIdCursorCodec.decodeOrBadRequest(cursor);
        List<Artist> bands = decoded == null
                ? artistRepository.findBandsPageFirst(q, pageable)
                : artistRepository.findBandsPageAfter(q, decoded.createdAt(), decoded.id(), pageable);

        boolean hasNext = bands.size() > safeLimit;
        List<Artist> slice = hasNext ? bands.subList(0, safeLimit) : bands;

        List<BandResponse> items = new ArrayList<>(slice.size());
        for (Artist band : slice) {
            items.add(toBandResponse(band, fetchAllSongsForBand(userId, band.getName())));
        }

        Artist last = slice.isEmpty() ? null : slice.getLast();
        String nextCursor = (hasNext && last != null && last.getCreatedAt() != null)
                ? LocalDateTimeIdCursorCodec.encode(last.getCreatedAt(), last.getId())
                : null;

        return new CursorPageResponse<>(items, nextCursor, hasNext);
    }

    public BandResponse getBand(UUID id) {
        Artist artist = artistRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Band not found: " + id));
        if (artist.getIsBand() == null || !artist.getIsBand()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Band not found: " + id);
        }
        return toBandResponse(artist, List.of());
    }

    private List<SongResponse> fetchAllSongsForBand(UUID userId, String bandName) {
        String q = bandName == null ? "" : bandName.trim();
        if (q.isBlank()) {
            return List.of();
        }

        List<SongResponse> out = new ArrayList<>();
        String cursor = null;
        int guard = 0;
        do {
            CursorPageResponse<SongResponse> page = searchService.searchSongs(userId, q, 100, cursor);
            out.addAll(page.items());
            cursor = page.nextCursor();
            guard++;
            if (guard > MAX_SEARCH_PAGES) {
                break;
            }
        } while (cursor != null);
        return out;
    }

    private static BandResponse toBandResponse(Artist artist, List<SongResponse> songs) {
        return BandResponse.builder()
                .id(artist.getId())
                .name(artist.getName())
                .sortName(artist.getSortName())
                .countryCode(artist.getCountryCode())
                .isBand(artist.getIsBand())
                .debutYear(artist.getDebutYear())
                .popularity(artist.getPopularity())
                .createdAt(artist.getCreatedAt())
                .updatedAt(artist.getUpdatedAt())
                .coverUrl(artist.getCoverUrl() == null ? null : artist.getCoverUrl().get("default"))
                .songs(songs)
                .build();
    }
}

File: src/main/kotlin/com/vjiki/music/service/BandService.kt
package com.vjiki.music.service

import com.vjiki.music.dto.BandResponse
import com.vjiki.music.dto.SongResponse
import com.vjiki.music.pagination.LocalDateTimeIdCursorCodec
import com.vjiki.music.pagination.CursorPageResponse
import com.vjiki.music.repository.ArtistRepository
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import java.util.UUID

@Service
class BandService(
    private val artistRepository: ArtistRepository,
    private val searchService: SearchService
) {
    fun getBands(limit: Int): List<BandResponse> {
        val safeLimit = limit.coerceIn(1, 200)
        return artistRepository.findBands(PageRequest.of(0, safeLimit)).map { it.toBandResponse() }
    }

    fun getBandsPage(userId: UUID, name: String, limit: Int, cursor: String?): CursorPageResponse<BandResponse> {
        val q = name.trim()
        if (q.isBlank()) throw ResponseStatusException(HttpStatus.BAD_REQUEST, "name is required")

        val safeLimit = limit.coerceIn(1, 200)
        val pageable = PageRequest.of(0, safeLimit + 1)

        val decoded = LocalDateTimeIdCursorCodec.decodeOrBadRequest(cursor)
        val bands = if (decoded == null) {
            artistRepository.findBandsPageFirst(q = q, pageable = pageable)
        } else {
            artistRepository.findBandsPageAfter(
                q = q,
                cursorCreatedAt = decoded.createdAt,
                cursorId = decoded.id,
                pageable = pageable
            )
        }

        val hasNext = bands.size > safeLimit
        val slice = if (hasNext) bands.take(safeLimit) else bands
        val items = slice.map { band ->
            val songs = fetchAllSongsForBand(userId = userId, bandName = band.name)
            band.toBandResponse(songs = songs)
        }

        val last = slice.lastOrNull()
        val nextCursor = if (hasNext && last?.createdAt != null) {
            LocalDateTimeIdCursorCodec.encode(last.createdAt, last.id)
        } else {
            null
        }

        return CursorPageResponse(items = items, nextCursor = nextCursor, hasNext = hasNext)
    }

    fun getBand(id: UUID): BandResponse {
        val artist = artistRepository.findById(id)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Band not found: $id") }
        if (!artist.isBand) throw ResponseStatusException(HttpStatus.NOT_FOUND, "Band not found: $id")
        return artist.toBandResponse()
    }

    private fun fetchAllSongsForBand(userId: UUID, bandName: String): List<SongResponse> {
        val q = bandName.trim()
        if (q.isBlank()) return emptyList()

        val out = mutableListOf<SongResponse>()
        var cursor: String? = null
        var guard = 0
        do {
            val page = searchService.searchSongs(userId = userId, q = q, limit = 100, cursor = cursor)
            out.addAll(page.items)
            cursor = page.nextCursor
            guard++
            if (guard > 50) break
        } while (cursor != null)
        return out
    }
}

private fun com.vjiki.music.entity.Artist.toBandResponse(songs: List<SongResponse> = emptyList()): BandResponse =
    BandResponse(
        id = id,
        name = name,
        sortName = sortName,
        countryCode = countryCode,
        isBand = isBand,
        debutYear = debutYear,
        popularity = popularity,
        createdAt = createdAt,
        updatedAt = updatedAt,
        coverUrl = coverUrl["default"],
        songs = songs
    )


